### PR TITLE
Slice sql

### LINKED ISSUE #1072: Slicing on Interactive Data with SQL does not work
Hi all,

Big fan of the project. I've started working on a 10 minutes to blaze tutorial (#423) and noticed that Slicing doesn't work for interactive objects. See [here](http://nbviewer.ipython.org/github/Will-So/Blaze-Workbook/blob/Slicing.ipynb) for an example.

I'd be happy to work on this but I still don't have a good grasp how exactly the `Data` function works. Any tips would be appreciated.

### PR DESCRIPTION
**Pull request by Will-So**
Closes #1072. I also added some docstrings as I was understanding how slicing works. 

```python
from blaze import Data
db = Data('sqlite:////' + DATA_DIR + 'lahman2013.sqlite')
df = db.Batting

import pandas as pd
from odo import odo
pandas_df = odo(df, pd.DataFrame)
```

<div class="cell code_cell unrendered selected" tabindex="2"><div class="input"><div class="prompt input_prompt">In&nbsp;[10]:</div><div class="inner_cell"><div class="ctb_hideshow"><div class="celltoolbar"></div></div><div class="input_area"><div class="CodeMirror cm-s-ipython"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 5.59375px; left: 89.59375px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" not-content="true" style="min-width: 18px;"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" not-content="true"></div><div class="CodeMirror-gutter-filler" not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; min-width: 112.2177734375px; margin-bottom: 0px; border-right-width: 30px; min-height: 28px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines"><div style="position: relative; outline: none;"><div class="CodeMirror-measure"></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors"><div class="CodeMirror-cursor" style="left: 84px; top: 0px; height: 17px;">&nbsp;</div></div><div class="CodeMirror-code"><pre class=""><span style="padding-right: 0.1px;"><span class="cm-variable">df</span>[<span class="cm-number">1002</span>:<span class="cm-number">1004</span>]</span></pre></div></div></div></div></div><div style="position: absolute; height: 30px; width: 1px; top: 28px;"></div><div class="CodeMirror-gutters" style="display: none; height: 58px;"></div></div></div></div></div></div><div class="widget-area" style="display: none;"><div class="prompt"><button class="close">×</button></div><div class="widget-subarea"></div></div><div class="output_wrapper"><div class="out_prompt_overlay prompt" title="click to scroll output; double click to hide" style="display: block;"></div><div class="output" style="display: flex;"><div class="output_area"><div class="prompt output_prompt">Out[10]:</div><div class="output_subarea output_html rendered_html output_result"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>playerID</th>
      <th>yearID</th>
      <th>stint</th>
      <th>teamID</th>
      <th>lgID</th>
      <th>G</th>
      <th>G_batting</th>
      <th>AB</th>
      <th>R</th>
      <th>H</th>
      <th>2B</th>
      <th>3B</th>
      <th>HR</th>
      <th>RBI</th>
      <th>SB</th>
      <th>CS</th>
      <th>BB</th>
      <th>SO</th>
      <th>IBB</th>
      <th>HBP</th>
      <th>SH</th>
      <th>SF</th>
      <th>GIDP</th>
      <th>G_old</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>allench01</td>
      <td>2002</td>
      <td>1</td>
      <td>CLE</td>
      <td>AL</td>
      <td>5</td>
      <td>5</td>
      <td>10</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>5</td>
    </tr>
    <tr>
      <th>1</th>
      <td>allench01</td>
      <td>2003</td>
      <td>1</td>
      <td>FLO</td>
      <td>NL</td>
      <td>12</td>
      <td>12</td>
      <td>24</td>
      <td>2</td>
      <td>5</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>5</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>12</td>
    </tr>
  </tbody>
</table></div></div></div><div class="btn btn-default output_collapsed" title="click to expand output" style="display: none;">. . .</div></div></div>

This will often have the same order as a Pandas DF even though order isn't guaranteed. 

<div class="cell code_cell selected unrendered" tabindex="2"><div class="input"><div class="prompt input_prompt">In&nbsp;[13]:</div><div class="inner_cell"><div class="ctb_hideshow"><div class="celltoolbar"></div></div><div class="input_area"><div class="CodeMirror cm-s-ipython CodeMirror-focused"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 5.59375px; left: 55.59375px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" not-content="true" style="min-width: 18px;"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" not-content="true"></div><div class="CodeMirror-gutter-filler" not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; min-width: 171.02734375px; margin-bottom: 0px; border-right-width: 30px; min-height: 28px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines"><div style="position: relative; outline: none;"><div class="CodeMirror-measure"></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors"><div class="CodeMirror-cursor" style="left: 50px; top: 0px; height: 17px;">&nbsp;</div></div><div class="CodeMirror-code"><pre class=""><span style="padding-right: 0.1px;"><span class="cm-variable">pandas_df</span>[<span class="cm-number">1002</span>:<span class="cm-number">1004</span>]</span></pre></div></div></div></div></div><div style="position: absolute; height: 30px; width: 1px; top: 28px;"></div><div class="CodeMirror-gutters" style="display: none; height: 58px;"></div></div></div></div></div></div><div class="widget-area" style="display: none;"><div class="prompt"><button class="close">×</button></div><div class="widget-subarea"></div></div><div class="output_wrapper"><div class="out_prompt_overlay prompt" title="click to scroll output; double click to hide" style=""></div><div class="output" style=""><div class="output_area"><div class="prompt output_prompt">Out[13]:</div><div class="output_subarea output_html rendered_html output_result"><div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>playerID</th>
      <th>yearID</th>
      <th>stint</th>
      <th>teamID</th>
      <th>lgID</th>
      <th>G</th>
      <th>G_batting</th>
      <th>AB</th>
      <th>R</th>
      <th>H</th>
      <th>...</th>
      <th>SB</th>
      <th>CS</th>
      <th>BB</th>
      <th>SO</th>
      <th>IBB</th>
      <th>HBP</th>
      <th>SH</th>
      <th>SF</th>
      <th>GIDP</th>
      <th>G_old</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1002</th>
      <td>allench01</td>
      <td>2002</td>
      <td>1</td>
      <td>CLE</td>
      <td>AL</td>
      <td>5</td>
      <td>5</td>
      <td>10</td>
      <td>0</td>
      <td>1</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>5</td>
    </tr>
    <tr>
      <th>1003</th>
      <td>allench01</td>
      <td>2003</td>
      <td>1</td>
      <td>FLO</td>
      <td>NL</td>
      <td>12</td>
      <td>12</td>
      <td>24</td>
      <td>2</td>
      <td>5</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>5</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>12</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 24 columns</p>
</div></div></div></div><div class="btn btn-default output_collapsed" title="click to expand output" style="display: none;">. . .</div></div></div>

Open ended slices will also work

<div class="cell code_cell rendered selected" tabindex="2"><div class="input"><div class="prompt input_prompt">In&nbsp;[12]:</div><div class="inner_cell"><div class="ctb_hideshow"><div class="celltoolbar"></div></div><div class="input_area"><div class="CodeMirror cm-s-ipython"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 5.59375px; left: 114.8115234375px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" not-content="true" style="min-width: 18px;"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" not-content="true"></div><div class="CodeMirror-gutter-filler" not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; min-width: 112.2177734375px; margin-bottom: 0px; border-right-width: 30px; min-height: 28px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines"><div style="position: relative; outline: none;"><div class="CodeMirror-measure"></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors" style=""><div class="CodeMirror-cursor" style="left: 109.2177734375px; top: 0px; height: 17px;">&nbsp;</div></div><div class="CodeMirror-code"><pre class=""><span style="padding-right: 0.1px;"><span class="cm-variable">df</span>[:].<span class="cm-variable">count</span><span class=" CodeMirror-matchingbracket">(</span><span class=" CodeMirror-matchingbracket">)</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 30px; width: 1px; top: 28px;"></div><div class="CodeMirror-gutters" style="display: none; height: 58px;"></div></div></div></div></div></div><div class="widget-area" style="display: none;"><div class="prompt"><button class="close">×</button></div><div class="widget-subarea"></div></div><div class="output_wrapper"><div class="out_prompt_overlay prompt" title="click to scroll output; double click to hide" style=""></div><div class="output" style=""><div class="output_area"><div class="prompt output_prompt">Out[12]:</div><div class="output_subarea output_html rendered_html output_result">97889</div></div></div><div class="btn btn-default output_collapsed" title="click to expand output" style="display: none;">. . .</div></div></div>

This operations would benefit from testing fixtures as proposed in #1026. 

I'm not quite sure why `replace_slices()` is returning a tuple of shape ((start, stop), ). Any insight on that would be helpful.

### CODE REVIEW

File: blaze/compute/sql.py [hunk 1]
```diff
@@ -789,3 +792,37 @@ def post_compute(_, s, **kwargs):
 @dispatch(IsIn, ColumnElement)
 def compute_up(expr, data, **kwargs):
     return data.in_(expr._keys)
+
+
+@dispatch(Slice, ClauseElement)
```
Comment (cpcloud) [inline blaze/compute/sql.py]: `ClauseElement` is too general here. It is the base class for *every* SQLAlchemy expression. We only want to slice the following things: `ColumnElement`s, `Selectable`s, and `Select` statements. `ColumnElement` includes things like `Column`s and any function on `Column`s. A `Selectable` is anything that can show up in the `from` clause in a `select` statement.

I'd dispatch like this:

```python
@dispatch(Slice, (ColumnElement, Selectable, Select))
```

File: blaze/compute/tests/test_sql_compute.py [hunk 2]
```diff
@@ -1495,3 +1495,19 @@ def test_do_not_erase_group_by_functions_with_datetime():
         extract(date from accdate.occurred_on)
     """
     assert normalize(result) == normalize(expected)
+
+
+def test_slice():
+    start, stop = 50, 100
```
Comment (cpcloud) [inline blaze/compute/tests/test_sql_compute.py]: Can you add another test that tests what happens when a step is passed in?

File: blaze/expr/expressions.py [hunk 3]
```diff
@@ -415,6 +415,17 @@ def sliceit(child, index):
 
 
 class Slice(Expr):
+    """
+    Elements `start` until `stop`
```
Comment (cpcloud) [inline blaze/expr/expressions.py]: `Slice` expressions also accept a `step` parameter on many backends. Please include that in the docstring here.

File: blaze/compute/sql.py [hunk 4]
```diff
@@ -789,3 +792,37 @@ def post_compute(_, s, **kwargs):
...
+        if start < 0:
+            raise ValueError('start value of slice cannot be negative'
+                             ' with a SQL backend')
+
+        stop = index.stop
+        if stop is not None and stop < 0:
+            raise ValueError('stop value of slice cannot be negative with a '
+                             'SQL backend.')
+
+        if index.step is not None:
```
Comment (cpcloud) [inline blaze/compute/sql.py]: We could allow `1` here, but that's not critical as `None` gives the same behavior.

File: blaze/expr/utils.py [hunk 5]
```diff
@@ -58,8 +58,21 @@ def hashable_index(index):
...
+
+    Examples
+    -------
+    >>> replace_slices([1,2,345,12])
+    [1, 2, 345, 12]
+    >>> type(replace_slices(_slice(1,5, None))) is slice
+    True
+    >>> type(replace_slices((2,5))) is tuple
+    True
+
```
Comment (cpcloud) [inline blaze/expr/utils.py]: Small style nit: can you put a space between the commas here?

File: blaze/compute/sql.py (hunk 1 above)
Comment (cpcloud) [inline blaze/compute/sql.py]: I like the idea of a warning when this is triggered. We should let people know what kind of assumption they're making when they choose to use `Slice` expressions on a relational database table. Something like:

```
The order of the result set from a Slice expression computed with the SQL backend is not deterministic.
```

**Comment by Will-So** [comment]
Made the changes. Under python 3.4 Travis threw an error I don't really understand:

`blaze/compute/tests/test_dask.py::test_compute /home/travis/build.sh: line 41:  2623 Aborted                 (core dumped) coverage run --include='blaze/*' $(which py.test)`

I can't redo the test without write access but I don't think it was related to the changes I made.

**Comment by cpcloud** [comment]
That's a concurrency bug that from `dask` that occasionally pops up. I'm considering disabling dask testing for now until it's resolved. You can trigger a new travis run by rehashing the commit at `HEAD` followed by a force push, like this:

```sh
git commit --amend -C HEAD && git push --force
```

**Comment by cpcloud** [comment]
@Will-So Can you merge master back into your local branch and fix the merge conflicts? Happy to help there if you need it.

**Comment by cpcloud** [comment]
The test that exposed the concurrency bug in dask is disabled now, so you shouldn't see failures related to that anymore

**Comment by cpcloud** [comment]
@Will-So Can you write a test in `test_sqlite.py` for the following edge case:

```python
t = symbol('t', 'var * {a: float64, b: string}')
result = t[:0]
```

**Comment by cpcloud** [comment]
Just want to make sure that this correctly executes and returns an empty table.

**Comment by Will-So** [comment]
@Will-So

I'm not sure how I can test this without testing it against a populated table.

```
str(compute(t[:0],s))`

would just return the SQL query with `:param_1, 2` rather than the actual values that are being sent. It can be shown by inspection that it will work though since :param_1 = 0 when null is fed to the backend and so we end up with limit - which returns an empty table. 

Can you think of a way to do this? Or maybe it's a good time to start working on test fixtures for the test module. I have some free time this week.

**Comment by cpcloud** [comment]
You can test against sqlite by using `odo`:

```python
from blaze import symbol
from odo import odo
df = pd.DataFrame([('a', 1), ('b', 2), ('c', 3)], columns=list('abc'))
t = odo(df, 'sqlite:///db.db::t')
expr = symbol('t', discover(t))
tm.assert_frame_equal(odo(expr[:2], set), odo(df.iloc[:2], set))
```

**Comment by cpcloud** [comment]
closing in favor of #1125
